### PR TITLE
Create .gitignore before trigger build

### LINKED ISSUE #206: Idea: put a .gitignore in the .elixir_ls folder
Just a small idea: when creating the .elixir_ls folder, add a .gitignore file inside it with this content:

```
*
```

Given that almost everyone uses Git and no one (presumably) wants to add their .elixir_ls folder to their Git history, it would save users from having to set up an ignore rule themselves.

(Copied from https://github.com/JakeBecker/elixir-ls/issues/193 )

### PR DESCRIPTION
**Pull request by wingyplus**
Create a .gitignore file by using project_dir in server state before
trigger build to make sure it create in the right place.

Close #206

### CODE REVIEW

File: apps/language_server/lib/language_server/server.ex [hunk 1]
```diff
@@ -789,4 +790,14 @@ defmodule ElixirLS.LanguageServer.Server do
...
     state
   end
+
+  defp create_gitignore(%{project_dir: project_dir} = state) do
+    with gitignore_path <- Path.join([project_dir, ".elixir_ls", ".gitignore"]),
+         :ok <- gitignore_path |> Path.dirname() |> File.mkdir(),
+         :ok <- File.write(gitignore_path, "*", [:read]) do
+      state
+    else
+      _ -> state
```
Comment (wingyplus) [inline apps/language_server/lib/language_server/server.ex:800]: I'm not sure what should we do when have a problems while creating a file.

Comment (axelson) [reply apps/language_server/lib/language_server/server.ex:800]: I think just logging a warning is good (using `JsonRpc.log_message(:warning, message)`)

File: apps/language_server/lib/language_server/server.ex [hunk 2]
```diff
@@ -789,4 +790,14 @@ defmodule ElixirLS.LanguageServer.Server do
   defp set_project_dir(state, _) do
     state
   end
+
+  defp create_gitignore(%{project_dir: project_dir} = state) do
```
Comment (axelson) [inline apps/language_server/lib/language_server/server.ex:794]: Can you change this function to not overwrite the `.gitignore` if it already exists (and don't error in that case either). My reasoning is that if a user has overwritten the `.gitignore` for some reason then we should not overwrite their changes.

File: apps/language_server/lib/language_server/server.ex [hunk 3]
```diff
@@ -789,4 +790,14 @@ defmodule ElixirLS.LanguageServer.Server do
...
+    with gitignore_path <- Path.join([project_dir, ".elixir_ls", ".gitignore"]),
+         :ok <- gitignore_path |> Path.dirname() |> File.mkdir_p(),
+         :ok <- File.write(gitignore_path, "*", [:read]) do
```
Comment (axelson) [inline apps/language_server/lib/language_server/server.ex:797]: Why use `:read` here? Wouldn't `:write` make more sense?

Comment (wingyplus) [reply apps/language_server/lib/language_server/server.ex:797]: @axelson My intention is `.gitignore` should not written by anyone except owner of this file. That's why I set it to `:read`.

Comment (wingyplus) [reply apps/language_server/lib/language_server/server.ex]: You're right. This should be `:write`

**Comment by wingyplus** [comment]
@axelson your suggestion was addressed. please kindly review.

**Comment by wingyplus** [comment]
I see test failed in dialyzer. Not sure why it happens.

https://travis-ci.org/github/elixir-lsp/elixir-ls/jobs/684546804

File: apps/language_server/lib/language_server/server.ex [hunk 4]
```diff
@@ -789,4 +790,25 @@ defmodule ElixirLS.LanguageServer.Server do
...
+
+  defp create_gitignore(%{project_dir: project_dir} = state) do
+    with gitignore_path <- Path.join([project_dir, ".elixir_ls", ".gitignore"]),
+         false <- File.exists?(gitignore_path),
+         :ok <- gitignore_path |> Path.dirname() |> File.mkdir_p(),
+         :ok <- File.write(gitignore_path, "*", [:write]) do
+      state
+    else
+      true ->
+        JsonRpc.log_message(:warning, ".elixir_ls/.gitignore already exists")
```
Comment (axelson) [inline apps/language_server/lib/language_server/server.ex:802]: Since this is an expected case, we don't need this log message.

**Comment by axelson** [comment]
I'm pretty sure that dialyzer error is spurious and unrelated.

File: apps/language_server/lib/language_server/server.ex (hunk 4 above)
Comment (wingyplus) [reply apps/language_server/lib/language_server/server.ex]: @axelson fixed.

**Comment by mikl** [comment]
@wingyplus yeah, thanks for taking the time to do this, it’ll be one less thing newcomers will have to deal with.

**Comment by wingyplus** [comment]
Thanks for you both. I learned a lot from this pr.

**Pull request merged**